src/metabase/models/permissions/parse.clj

```diff
@@ -33,34 +41,56 @@
+  [path-or-paths x]
+  (if (seqable? (first path-or-paths))
+    (map (fn [path] (append-to-all path x)) (seq path-or-paths))
+    (into path-or-paths [x])))
+
 (defn- path
   "Recursively build permission path from parse tree"
   [tree]
   (match/match tree
-    (_ :guard insta/failure?)    (log/error (trs "Error parsing permissions tree {0}" (pr-str tree)))
-    [:permission t]              (path t)
-    [:all]                       [:all] ; admin permissions
-    [:db db-id]                  (let [db-id (Long/parseUnsignedLong db-id)]
-                                   [[:db db-id :data :native :write]
-                                    [:db db-id :data :schemas :all]])
-    [:db db-id db-node]          (let [db-id (Long/parseUnsignedLong db-id)]
-                                   (into [:db db-id] (path db-node)))
-    [:schemas]                   [:data :schemas :all]
-    [:schemas schema]            (into [:data :schemas] (path schema))
-    [:schema schema-name]        [schema-name :all]
-    [:schema schema-name table]  (into [schema-name] (path table))
-    [:table table-id]            [(Long/parseUnsignedLong table-id) :all]
-    [:table table-id table-perm] (into [(Long/parseUnsignedLong table-id)] (path table-perm))
-    [:table-perm perm]            (case perm
-                                    "read"            [:read :all]
-                                    "query"           [:query :all]
-                                    "query/segmented" [:query :segmented])
-    [:native]                    [:data :native :write]
+    (_ :guard insta/failure?)      (log/error (trs "Error parsing permissions tree {0}" (pr-str tree)))
+    [:permission t]                (path t)
+    [:all]                         [:all] ; admin permissions
+    [:db db-id]                    (let [db-id (Long/parseUnsignedLong db-id)]
+                                     [[:db db-id :data :native :write]
+                                      [:db db-id :data :schemas :all]])
+    [:db db-id db-node]            (let [db-id (Long/parseUnsignedLong db-id)]
+                                     (into [:db db-id] (path db-node)))
+    [:schemas]                     [:data :schemas :all]
+    [:schemas schema]              (into [:data :schemas] (path schema))
+    [:schema schema-name]          [schema-name :all]
+    [:schema schema-name table]    (into [schema-name] (path table))
+    [:table table-id]              [(Long/parseUnsignedLong table-id) :all]
+    [:table table-id table-perm]   (into [(Long/parseUnsignedLong table-id)] (path table-perm))
+    [:table-perm perm]              (case perm
+                                      "read"            [:read :all]
+                                      "query"           [:query :all]
+                                      "query/segmented" [:query :segmented])
+    [:native]                      [:data :native :write]
+    ;; download perms
```

Comment:
The only added lines in this function are the download-related ones directly under this comment. The rest is just whitespace changes.